### ISSUE TITLE
Unexpected disconnect when server sends connection: close

### ISSUE THREAD

**Issue by Falmarri** [akka/akka#20570]
```
Caused by: akka.http.impl.engine.client.PoolSlot$UnexpectedDisconnectException: Unexpected disconnect
	at akka.http.impl.engine.client.PoolSlot$SlotProcessor$$anonfun$7$$anonfun$8.apply(PoolSlot.scala:183) ~[akka-http-core_2.11-2.4.5.jar:na]
```

This is the code

```scala
import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.client.RequestBuilding._
import akka.http.scaladsl.model.headers.Connection
import akka.http.scaladsl.model.{HttpResponse, StatusCodes}
import akka.http.scaladsl.unmarshalling.Unmarshal
import akka.stream.scaladsl.{Sink, Source}
import akka.stream.{ActorMaterializer, ActorMaterializerSettings}
import com.typesafe.config.ConfigFactory

import scala.concurrent.Await
import scala.concurrent.duration._


object Main2{

    val config = ConfigFactory.load()

    {
        implicit val system = ActorSystem("root")
        implicit val executor = system.dispatcher
        implicit val materializer = ActorMaterializer(ActorMaterializerSettings(system).withFuzzing(true))

        val s = Await.result(Http().bindAndHandleSync(_ => HttpResponse(StatusCodes.OK, Connection("close") :: Nil).withEntity(scala.util.Random.alphanumeric.take(21000).mkString), "0.0.0.0", 1234), Duration.Inf)
    }

    def main(args: Array[String]) {
        implicit val system = ActorSystem("root2")
        implicit val executor = system.dispatcher
        implicit val materializer = ActorMaterializer(ActorMaterializerSettings(system).withFuzzing(true))
        val l = Source(1 to 1000).mapAsync(1){ r =>
            println(r)
            Http().singleRequest(Post("http://localhost:1234"))

        }
        .mapAsync(1){r => Unmarshal(r.entity).to[String]}
        .map(v => println(v.length))
        .runWith(Sink.ignore)

        val r = Await.ready(l, 1 minute)
        println(r)
    }
}
```

**Comment by 2beaucoup** [comment]
What's the outcome? How many responses do you actually receive?

**Comment by 2beaucoup** [comment]
Do you have additional settings configured (like i.e. pipelining)?

**Comment by Falmarri** [comment]
No additional settings. Outcome is an excetion:

```
Caused by: akka.http.impl.engine.client.PoolSlot$UnexpectedDisconnectException: Unexpected disconnect
    at akka.http.impl.engine.client.PoolSlot$SlotProcessor$$anonfun$7$$anonfun$8.apply(PoolSlot.scala:183) ~[akka-http-core_2.11-2.4.5.jar:na]
```

Usually after only 4-5 requests, but it's not predictable.

**Comment by 2beaucoup** [comment]
Try to increase `host-connection-pool.max-connections` . Replies with `Connection: close` close the connection so they cannot be reused.

**Comment by Falmarri** [comment]
I understand that they close the connection. Shouldn't akka be responsible for maintaining its connection pool in a state that is usable? Are you saying that my code as it exists above shouldn't be expected to work?

**Comment by Falmarri** [comment]
Also, the same issue happens no matter what the max-connections size is (although I haven't tried setting it > 1000)

**Comment by Joe-Edwards** [comment]
I think this is related - if not I'll raise a separate ticket.

I've recently been running a test which sends several requests (using the default connection pool) where the response contains a 'Connection: close' (caused by an early response, but I don't think that's important).

Even if I put a _10 second_ sleep between each request (giving the connection pool plenty of time to discard the closed connections) I still see >50% failing with a connection failure.

I understand that pipelined requests will fail when the connection is killed, but I'd expect a better success rate for requests this far apart.

**Comment by ktoso** [comment]
"This far apart" does not quite describe the situation though, since there are internal buffers between stages requests may be in buffers when the connection gets closed. 

Issue worth investigating in any case.

**Comment by ktoso** [comment]
I'm not convinced it's a "bug", but want to work on it soon to determine that, and if we can do better than currently.

**Comment by Falmarri** [comment]
@ktoso how could this not be a bug? So you're saying that if a server sends a connection: close, we shouldn't be expected to be able to reliably consume the entity?

**Comment by ktoso** [comment]
Only thing I'm saying is I haven't had the time to actually sit down and focus on this issue, so I have to do that. I marked it as a bug do to exactly that ;)

**Comment by Falmarri** [comment]
@ktoso gotcha

**Comment by 2beaucoup** [comment]
This test added to `akka.http.impl.engine.client.ConnectionPoolSpec` succeeds:
```scala

"be able to handle 500 `Connection: close` requests against the test server" in new TestSetup {
  val settings = ConnectionPoolSettings(system).withMaxConnections(4)
  val poolFlow = Http().cachedHostConnectionPool[Int](serverHostName, serverPort, settings = settings)

  val N = 500
  val requestIds = Source.fromIterator(() ⇒ Iterator.from(1)).take(N)
  val idSum = requestIds.map(id ⇒ HttpRequest(uri = s"/r$id").withHeaders(Connection("close")) -> id).via(poolFlow).map {
    case (Success(response), id) ⇒
      requestUri(response) should endWith(s"/r$id")
      id
    case x ⇒ fail(x.toString)
  }.runFold(0)(_ + _)

  (1 to N).foreach(_ ⇒ acceptIncomingConnection())

  Await.result(idSum, 10.seconds) shouldEqual N * (N + 1) / 2
}
```
Output: https://gist.github.com/2beaucoup/f5a70733dedadf17a1f425287370a73f

**Comment by Joe-Edwards** [comment]
My minimal example looks like: https://gist.github.com/Joe-Edwards/f5d8c067997ad56cfe6b209b0ef0a598

The output varies, but something like

```
Failure(akka.stream.StreamTcpException: The connection closed with error: An established connection was aborted by the software in your host machine)
Failure(akka.stream.StreamTcpException: The connection closed with error: An established connection was aborted by the software in your host machine)
Failure(akka.stream.StreamTcpException: The connection closed with error: An established connection was aborted by the software in your host machine)
Success(HttpResponse(400 Bad Request,List(Server: akka-http/2.4.5, Date: Tue, 24 May 2016 11:00:12 GMT, Connection: close),HttpEntity.Strict(text/plain; charset=UTF-8,test response),HttpProtocol(HTTP/1.1)))
Failure(akka.stream.StreamTcpException: The connection closed with error: An established connection was aborted by the software in your host machine)
Success(HttpResponse(400 Bad Request,List(Server: akka-http/2.4.5, Date: Tue, 24 May 2016 11:00:14 GMT, Connection: close),HttpEntity.Strict(text/plain; charset=UTF-8,test response),HttpProtocol(HTTP/1.1)))
Success(HttpResponse(400 Bad Request,List(Server: akka-http/2.4.5, Date: Tue, 24 May 2016 11:00:15 GMT, Connection: close),HttpEntity.Strict(text/plain; charset=UTF-8,test response),HttpProtocol(HTTP/1.1)))
Failure(akka.stream.StreamTcpException: The connection closed with error: An established connection was aborted by the software in your host machine)
Success(HttpResponse(400 Bad Request,List(Server: akka-http/2.4.5, Date: Tue, 24 May 2016 11:00:17 GMT, Connection: close),HttpEntity.Strict(text/plain; charset=UTF-8,test response),HttpProtocol(HTTP/1.1)))
Failure(akka.stream.StreamTcpException: The connection closed with error: An established connection was aborted by the software in your host machine)
```
Particularly weird is that the _first_ request doesn't succeed - suggesting it's nothing to do with connection recycling. I haven't investigated whether triggering the connection close by other means causes the same behaviour.

**Comment by cppexpert** [comment]
This happens to me too along with lots of 

```
java.lang.IllegalStateException: The first message must be `ExposedPublisher` but was [FromConnection(OnComplete)]
	at akka.http.impl.engine.client.PoolSlot$SlotProcessor$$anonfun$waitingExposedPublisher$1.applyOrElse(PoolSlot.scala:106)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:482)
	at akka.http.impl.engine.client.PoolSlot$SlotProcessor.akka$stream$actor$ActorSubscriber$$super$aroundReceive(PoolSlot.scala:89)
	at akka.stream.actor.ActorSubscriber$class.aroundReceive(ActorSubscriber.scala:201)
	at akka.http.impl.engine.client.PoolSlot$SlotProcessor.akka$stream$actor$ActorPublisher$$super$aroundReceive(PoolSlot.scala:89)
	at akka.stream.actor.ActorPublisher$class.aroundReceive(ActorPublisher.scala:326)
	at akka.http.impl.engine.client.PoolSlot$SlotProcessor.aroundReceive(PoolSlot.scala:89)
```

**Comment by viktorklang** [comment]
@cppexpert When someone has fixed it… (PRs are welcome)

**Comment by Falmarri** [comment]
Any tips on debugging this?  I tried and couldn't work out what was happening. My guess is that in OutgoingConnectionBlueprint, the Termination Backchannel part of the flow is completing before the response can be parsed and merged. But I can't figure out how to prove/disprove that.

**Comment by kodemaniak** [comment]
I get the same error when I access my backend service with an akka-http client via a (nginx) proxy. I had a very similar problem 3 years ago with spray. Nginx forces a connection: close on keep-alive connections every 100 requests or so. Spray didn't handle that appropriately. Could this be a similar problem? If I run my code against the backend service directly, I don't get the exception.

**Comment by kodemaniak** [comment]
I have refactored the reproducer that I wrote for spray three years ago to use akka-http instead, and it reproduces the error: https://github.com/kodemaniak/akka-http-connection-close

If you start 'MyService' and then run the Client, after a while the exception should be thrown. If you do the same number of requests against 'MyService' with ab, e.g., it works.

**Comment by 2beaucoup** [comment]
The issue here was that the `PoolSlot` was still sending requests after a `Connection: close` response was received but the corresponding `Complete` didn't arrive yet.

#20816 fixes that.

**Comment by Falmarri** [comment]
<3 I looked at that code for so long but didn't think about adding a new context state.

**Comment by cppexpert** [comment]
@2beaucoup What version do we expect fix to be in ?

**Comment by 2beaucoup** [comment]
> <3 I looked at that code for so long but didn't think about adding a new context state.

@Falmarri Me too! Maybe you didn't look long enough? :wink:

Would be awesome if anyone affected by this could verify that this actually fixes things.

**Comment by 2beaucoup** [comment]
A solution would be to refactor the `Poolflow` into a `GraphStage`, wrap the connection flows into `SubSourceOutlet`s/`SubSinkInlet`s and remove `PoolConductor`/`PoolSlot`. I followed that route but it looks like such a subflow can only have one inflight element, which makes pipelining impossible. I'll post a test case later.

We could even move over and use https://github.com/akka/akka-stream-contrib/pull/26 for the retry logic.

/cc @rkuhn @sirthias

**Comment by drewhk** [comment]
As for the SubSource/Sink stuff, they are currently one-by-one but you can add a buffer on the target side and also, we can add a buffer into the SubThings itself if needed. Plus, they are just a wrapper around AsyncCallback, so you can build whatever else you want on top of that.

**Comment by ktoso** [comment]
@yakticus prepared a reproducer:

> All the details including reproducing the error are here: https://github.com/yakticus/akka-UnexpectedDisconnectException

**Comment by 2beaucoup** [comment]
I prepared a test case. Will post later.

**Comment by ktoso** [comment]
@2beaucoup did you have time to post the reproducer? I'll have time tomorrow to work on this so repro would help in that :)

**Comment by 2beaucoup** [comment]
@ktoso Sure. This switches between GET and POST and randomly inserts `Connection: close` (in ConnectionPoolSpec):
```scala
    "be able to handle 500 pipelined requests with connection termination" ignore new TestSetup(autoAccept = true) {
      def closeHeader() =
        if (util.Random.nextInt(8) == 0) Connection("close") :: Nil else Nil

      override def testServerHandler(connNr: Int): HttpRequest ⇒ HttpResponse = {
        case r: HttpRequest ⇒
          val idx = r.uri.path.tail.head.toString
          HttpResponse()
            .withHeaders(RawHeader("Req-Idx", idx) +: responseHeaders(r, connNr))
            .withDefaultHeaders(closeHeader())
      }

      for (pipeliningLimit ← Iterator.from(1).map(math.pow(2, _).toInt).take(4)) {
        val settings = ConnectionPoolSettings(system).withMaxConnections(4).withPipeliningLimit(pipeliningLimit)
        val poolFlow = Http().cachedHostConnectionPool[Int](serverHostName, serverPort, settings = settings)

        def method() =
          if (false && util.Random.nextInt(2) == 0) HttpMethods.POST else HttpMethods.GET

        def request(i: Int) =
          HttpRequest(method = method(), headers = closeHeader(), uri = s"/$i") → i

        val N = 500
        val idSum =
          Source.fromIterator(() ⇒ Iterator.from(1))
            .take(N)
            .map(request)
            .via(poolFlow)
            .map {
              case (Success(response), id) ⇒
                requestUri(response) should endWith(s"/$id")
                id
              case x ⇒ fail(x.toString)
            }.runFold(0)(_ + _)

        Await.result(idSum, 30.seconds) shouldEqual N * (N + 1) / 2
      }
    }
```

**Comment by jrudolph** [comment]
We hope that this problem is now fixed. The fix will be delivered with 2.4.10. Whoever was affected, please retry once 2.4.10 is released (due this week).

We know that there are still outstanding bugs related to pooling and pipelining (#21368) so we'll try to make incremental updates to improve the state of the client side implementation.
